R/ddl.R

```diff
@@ -0,0 +1,210 @@
+#'  shouldn't be specified when `code` is specified.
+#'
+#' @param code (`character`, `language`)\cr
+#'  Object containing code to be evaluated to load data. Shouldn't be specified when `expr`
+#'  is specified.
+#'
+#'
+#' @param ui (`shiny.tag`)\cr
+#'   `shiny` ui module containing inputs which `id` correspond to the
+#'   args in the `code`.
+#'
+#' @param server (`function`)\cr
+#'   `shiny` server module returning data. This server should execute
+#'   `code` and return a reactive data containing necessary data. To handle
+#'   evaluation and code masking process it is recommended to use `ddl_run`.
+#'   Package provides universal `username_password_server` which
+#'   runs `ddl_run` function, which returns `teal_data` object.
+#'   Details in the the example
+#'
+#' @param input_mask (`list` named)\cr
+#'   arguments to be substituted in the `code`. These
+#'   argument are going to replace arguments set through
+#'   `ui` and `server`. Example use case is when app user
+#'   is asked to input a password and we'd like to skip this
+#'   input in the reproducible code. Typically users password
+#'   is substituted with `askpass::askpass()` call, so the
+#'   returned code is still executable but secure.
+#'
+#' @param datanames (`character`)\cr
+#'   Names of the objects to be created from the code evaluation.
+#'   If not specified (`character(0)`), all objects will be used to `teal_data` function
+#'   (via `env_list` in `postprocess_fun`).
+#'
+#' @inheritParams teal.data::teal_data
+#'
+#'
+#' @export
+ddl <- function(expr,
+                code,
+                ui = submit_button_ui,
+                input_mask = list(),
+                server = submit_button_server,
+                join_keys = teal.data::join_keys(),
+                datanames = names(join_keys$get())) {
+  if (!missing(expr) && !missing(code)) {
+    stop("Only one of `expr` or `code` should be specified")
+  }
+  if (!missing(expr)) {
+    code <- substitute(expr)
+  }
+  if (is.character(code)) {
+    code <- parse(text = code)
+  }
+
+  if (length(datanames) == 0) {
+    stop("`datanames` argument is required")
+  }
+
+  ddl_object <- structure(
+    list(ui = ui, server = server),
+    code = code,
+    input_mask = input_mask,
+    datanames = datanames,
+    join_keys = join_keys,
+    class = "ddl"
+  )
+
+  # function defined here to have access to this environment
+  # ddl_run is a convenience wrapper over ddl_run2 to avoid specifying `x` argument
+  # thanks to this in the server one can simply call
+  #  `ddl_run(input = input)`
+  # instead of
+  #  `ddl_run(x = ddl_object, input = input))`
+  ddl_run <- function(input = list()) {
+    ddl_run2(x = ddl_object, input = input)
+  }
+
+  # changing enclosing environment of the server to have access to ddl_fun function
+  # Thanks to this ddl object contains only ui and server functions
+  #  and server function can be run just by calling ddl$server("<id>")!
```

Comment:
And pray people don't bind ddl_run in their server functions, huh? ;)

Reply:
I don't think this change is what you really want because people might in their server functions reference variables present in the enclosing environment of `server`. If you change the environment, you sever that connection. A smarter idea is to either:
* add `ddl_run` to the present enclosing environment of `server`
* insert another environment in the tree that contains only the `ddl_run` function. The environment tree would look like: ddl_run_env <- the old enclosing environment. I haven't tried, but it should work.

Reply:
I also guess that the current proposition has some chances of working in an interactive session, due to how environment() is going to have global env as an ancestor, so the values defined outside of server can be found anyway, but it really is going to break if people want to use a server function defined in a separate package, because the enclosing environment of a package function is usually "older" than the global env.

Reply:
Thanks for highlighting this! This is a very good remark and we will definitely rediscuss this.